test/common/matcher/matcher_test.cc

```
@@ -643,5 +643,172 @@ TEST_F(MatcherTest, RecursiveMatcherCannotMatch) {
+  top_matcher->addMatcher(createSingleMatcher("string", [](auto) { return true; }), on_match_2);
+
+  // Expect all first nested matches to be skipped and recorded due to keep_matching in parent.
+  ReenterableMatchEvaluator<TestData> reenterable_matcher(top_matcher, false);
+  std::vector<ActionFactoryCb> skipped_results;
+  MaybeMatchResult result_1 = reenterable_matcher.evaluateMatch(TestData(), skipped_results);
+  EXPECT_EQ(result_1.match_state_, MatchState::MatchComplete);
+  ASSERT_NE(result_1.result_, nullptr);
+  EXPECT_EQ(result_1.result_().get()->getTyped<StringAction>().string_, "match 3");
+  std::vector<std::string> expected_skipped_results = {"skipped - keep matching 1",
+                                                       "skipped - match 2"};
+  ASSERT_EQ(skipped_results.size(), expected_skipped_results.size());
+  for (size_t i = 0; i < skipped_results.size(); ++i) {
+    EXPECT_EQ(skipped_results.at(i)()->getTyped<StringAction>().string_,
+              expected_skipped_results[i]);
+  }
+  skipped_results.clear();
+
+  // Expect only the keep_matching nested matcher to be skipped from the second parent.
+  MaybeMatchResult result_2 = reenterable_matcher.evaluateMatch(TestData(), skipped_results);
+  EXPECT_EQ(result_2.match_state_, MatchState::MatchComplete);
+  ASSERT_NE(result_2.result_, nullptr);
+  EXPECT_EQ(result_2.result_().get()->getTyped<StringAction>().string_, "match 4");
+  ASSERT_EQ(skipped_results.size(), 1);
+  EXPECT_EQ(skipped_results.at(0)()->getTyped<StringAction>().string_, "keep matching 2");
+  skipped_results.clear();
+
+  MaybeMatchResult result_3 = reenterable_matcher.evaluateMatch(TestData(), skipped_results);
+  EXPECT_EQ(result_3.match_state_, MatchState::MatchComplete);
+  ASSERT_NE(result_3.result_, nullptr);
+  EXPECT_EQ(result_3.result_().get()->getTyped<StringAction>().string_, "on no match");
+  EXPECT_TRUE(skipped_results.empty());
```

Some test suggestions - if you merge main, I've added some matchers at the top of this file which should make this sort of thing less crash-prone and more readable, e.g. now you can replace
```
  ASSERT_NE(result_3.result_, nullptr);
  EXPECT_EQ(result_3.result_().get()->getTyped<StringAction>().string_, "on no match");
```
with
```
EXPECT_THAT(result_3.result_, IsStringAction("on no match"));
```

And then I didn't add a helper for this, but I would highly recommend
```
using testing::IsEmpty;
EXPECT_THAT(skipped_results, IsEmpty());
```
over `EXPECT_TRUE(skipped_results.empty())`, because then in the case of a test failure you get some information about what was in the list, rather than the information being only "it wasn't empty".
(In this case it won't help much because you'd need a custom printer for the contained types to get *really* useful output, but it's a step in the right direction.)

---

Updated